hack/run-hive-locally.sh

```diff
@@ -0,0 +1,349 @@
+#!/usr/bin/env bash
+
+# TODO: hiveadmission support
+# TODO: run hive-controllers/hive-clustersync separately with multiple replicas each (need docker to avoid port collisions)
+
+set -e
+set -u
+set -o pipefail
+
+# Highlighted echo
+function hl_echo {
+    tput setaf 3
+    echo "$@"
+    tput sgr0
+}
+
+function usage {
+    echo "The script runs a Hive component ($hive_component_operator or $hive_component_controllers) locally
+against the active cluster, simulating in-cluster environment. It records the in-cluster Hive
+configurations, scales down the selected component, then runs it locally with the same environment.
+On pressing ctrl + c, the script is terminated and relevant components are scaled back up.
+
+Warnings:
+- The active cluster may be in use, in which case the script could interfere with its functioning.
+
+Notes:
+- Artifacts (ConfigMaps containing in-cluster Hive configurations and log files) produced are stored in
+the --artifacts_folder which is removed once the script ends unless --preserve-artifacts is specified.
+- Selecting $hive_component_controllers runs all Hive controllers (including hive-clustersync) together
+as a single replica.
+- Endpoints of the Hive component running locally (e.g. metrics, pprof) are directly exposed on localhost.
+
+Usage:
+- ./hack/run-hive-locally.sh COMPONENT [--flag=value ...]
+
+Parameters:
+- COMPONENT: must be either $hive_component_operator or $hive_component_controllers
+
+Optional flags:
+- --hive-operator-ns=hive: namespace where the hive-operator is installed
+- --log-level=debug: log level of the component to run
+- --artifacts-folder=artifacts: folder that will be created to store the artifacts
+- --preserve-artifacts: if specified, preserve the artifacts folder after the script ends
+- --help"
+}
+
+function kill_background_jobs {
+    local pid
+    while read -r pid; do
+        hl_echo "Killing process $pid"
+        kill "$pid" || true
+    done < <(jobs -p)
+    wait
```

Comment:
This might be too much detail here but I'd like to jot down some thoughts. 

`jobs -p` returns PID of the background job's process group leader. So in case of `sleep 180 | sleep 180 | sleep 180 & `, this function only kills the first sleep process, not all of them. In our case, however, this function kills the `make run` (or `./bin/manager` depending on what iteration of the while loop it is), causing the stdin to the `tee` command to be closed, so `tee` terminates once all data it has are flushed to its outputs. In other words, this function is imperfect but seem to work in our case. 

I once tried to kill the job number instead, something like `kill %n`. This does seem to work, killing all processes which belong to the job / process group. Unfortunately, it looks like something in the `make run | tee` pipeline does not handle SIGTERM properly, resulting in `make run` ending abruptly. This messes with Hive's leader election process, as the leader is killed before releasing the lease, forcing the next deployment (after HiveConfig.spec changes) to wait for a few minutes.